### DISCUSSION TITLE
Typescript 5.0 decorator support

### DISCUSSION THREAD

**Discussion by mikehall314** [mikro-orm/mikro-orm#4148] [General]
TypeScript 5.0 has been announced, with breaking changes to the Decorators API.

https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/

Are there plans to update MikroORM to support the new decorators?

**Comment by B4nan** [comment]
MikroORM should be already compatible, it already uses TS 5.0. 

I wouldn't call this "breaking changes to decorators", they added support for ES decorators, while they keep supporting the CJS decorators too, just like before. Note that your project would have to be ESM to use those new decorator rules. And lastly, keep in mind ES decorators don't support metadata, so they are a bit useless here, e.g. you will need to declare every single property type including scalars, no more `reflect-metadata` support.

**Reply by ml1nk** [reply]
In my understanding if you enable experimentalDecorators in your tsconfig, you can't use ES decorators in your code. If experimentalDecorators is disabled, then the "old" Decorators don't work. I haven't found a way to write meaningful decorators that works in both worlds. 

experimentalDecorators: true
![image](https://github.com/mikro-orm/mikro-orm/assets/4943440/b42f8a41-b715-4a66-9fae-da7f155e74d0)

experimentalDecorators: false
![image](https://github.com/mikro-orm/mikro-orm/assets/4943440/bec2d824-9b6a-4298-aa37-1bea0cb49530)

@B4nan 
As such you can't use MikroORM with the ES Decorators in the same project or am I wrong?

> And lastly, keep in mind ES decorators don't support metadata, so they are a bit useless here, e.g. you will need to declare every single property type including scalars, no more reflect-metadata support.

Do you think ts-morph will still work like before if there are ES Decorators instead of the current ones?

**Reply by B4nan** [reply]
Yeah, that's quite possible, I don't think there is a way to have both at the same time. And given there is no type inference available in the ES decorators, I will stick with the old definition for now.

Maybe we could have a separate export for ES decorator variants?

> Do you think ts-morph will still work like before if there are ES Decorators instead of the current ones?

Yes, as ts-morph is not about decorators, it extracts the type information connected to the class property, the decorator above it is irrelevant.

**Reply by B4nan** [reply]
I did a few tests and I think we could maybe find a way to support both signatures, with help of few `any` types, not great but better than failing with an error like this.

**Reply by ml1nk** [reply]
Separate exports with ES decorators or compatiblity by smart typing are both valid in my opinion.

**Reply by ZirionNeft** [reply]
Any news on this?

**Reply by piotr-cz** [reply]
The blogpost TypeScript 5.0 > [Differences with Experimental Legacy Decorators](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#differences-with-experimental-legacy-decorators) states that new (TC39 Stage 3) decorators added in TypeScript 5.0 

> This new decorators proposal is not compatible with `--emitDecoratorMetadata`, and it does not allow **decorating parameters**.
> Future ECMAScript proposals may be able to help bridge that gap.

So it would be not possible yet to use them for [MikroORM Entity Properties](https://mikro-orm.io/docs/decorators#entity-properties) like `@Property()`, `@PrimaryKey()`, `@Enum()` and so on

**Reply by B4nan** [reply]
_parameters_ are not _properties_, you can use ES decorators just fine for what we are doing here, we don't use parameter decorators at all. the bigger issue is that there is no support for reflection, so all types need to be specified explicitly (or you would need tsmorph). v7 will most likely use ES decorators (the old spec will be still available too)
